drm-glesv2 compile failure with meson build-system

I'm cross-compiling for `drm-glesv2` which has worked fine under the older waf build system for aeons. I've just bumped version and compile now fails after "Build: Meson build system support" (git bisect breaking commit). The OS is LibreELEC which uses DRM/GBM to run Kodi. The build log shows it including X11 things. Looking at `eglplatform.h` I suspect it's not detecting `__GBM__` correctly?

```
BUILD      glmark2 (target)
    TOOLCHAIN      meson (auto-detect)
Executing (target): meson --prefix=/usr --bindir=/usr/bin --sbindir=/usr/sbin --sysconfdir=/etc --libdir=/usr/lib --libexecdir=/usr/lib --localstatedir=/var --buildtype=plain -Dstrip=true --cross-file=/home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/build/glmark2-877cc5a3d6e34c18b5735a42f7b47e7ec92b0dd8/.armv8a-libreelec-linux-gnueabihf/meson.conf -Dflavors=drm-glesv2 /home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/build/glmark2-877cc5a3d6e34c18b5735a42f7b47e7ec92b0dd8
DEPRECATION: c_args in the [properties] section of the machine file is deprecated, use the [built-in options] section.
DEPRECATION: c_link_args in the [properties] section of the machine file is deprecated, use the [built-in options] section.
DEPRECATION: cpp_args in the [properties] section of the machine file is deprecated, use the [built-in options] section.
DEPRECATION: cpp_link_args in the [properties] section of the machine file is deprecated, use the [built-in options] section.
Using 'PKG_CONFIG_PATH' from environment with value: ''
The Meson build system
Version: 0.56.0
Source dir: /home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/build/glmark2-877cc5a3d6e34c18b5735a42f7b47e7ec92b0dd8
Build dir: /home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/build/glmark2-877cc5a3d6e34c18b5735a42f7b47e7ec92b0dd8/.armv8a-libreelec-linux-gnueabihf
Build type: cross build
Project name: glmark2
Project version: 2020.04
Using 'CFLAGS' from environment with value: '-march=native -O2 -Wall -pipe -I/home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/toolchain/include -Wno-format-security'
Using 'LDFLAGS' from environment with value: '-Wl,-rpath,/home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/toolchain/lib -L/home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/toolchain/lib'
Using 'CPPFLAGS' from environment with value: ''
C compiler for the host machine: /home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/toolchain/bin/armv8a-libreelec-linux-gnueabihf-gcc (gcc 10.2.0 "armv8a-libreelec-linux-gnueabihf-gcc-10.2.0 (GCC) 10.2.0")
C linker for the host machine: /home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/toolchain/bin/armv8a-libreelec-linux-gnueabihf-gcc ld.bfd 2.35.1
Using 'CXXFLAGS' from environment with value: '-march=native -O2 -Wall -pipe -I/home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/toolchain/include -Wno-format-security'
Using 'LDFLAGS' from environment with value: '-Wl,-rpath,/home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/toolchain/lib -L/home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/toolchain/lib'
Using 'CPPFLAGS' from environment with value: ''
C++ compiler for the host machine: /home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/toolchain/bin/armv8a-libreelec-linux-gnueabihf-g++ (gcc 10.2.0 "armv8a-libreelec-linux-gnueabihf-g++-10.2.0 (GCC) 10.2.0")
C++ linker for the host machine: /home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/toolchain/bin/armv8a-libreelec-linux-gnueabihf-g++ ld.bfd 2.35.1
Using 'CC_FOR_BUILD' from environment with value: '/home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/toolchain/bin/host-gcc'
C compiler for the build machine: /home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/toolchain/bin/host-gcc (gcc 7.5.0 "gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0")
C linker for the build machine: /home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/toolchain/bin/host-gcc ld.bfd 2.30
Using 'CXX_FOR_BUILD' from environment with value: '/home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/toolchain/bin/host-g++'
C++ compiler for the build machine: /home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/toolchain/bin/host-g++ (gcc 7.5.0 "g++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0")
C++ linker for the build machine: /home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/toolchain/bin/host-g++ ld.bfd 2.30
Build machine cpu family: x86_64
Build machine cpu: x86_64
Host machine cpu family: x86_64
Host machine cpu: x86_64
Target machine cpu family: arm
Target machine cpu: armv8-a
Library m found: YES
Library dl found: YES
Found pkg-config: /home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/toolchain/bin/pkg-config (0.29.2)
Using 'PKG_CONFIG_PATH' from environment with value: ''
Run-time dependency libjpeg found: YES 2.0.6
Using 'PKG_CONFIG_PATH' from environment with value: ''
Run-time dependency libpng16 found: YES 1.6.37
Using 'PKG_CONFIG_PATH' from environment with value: ''
Run-time dependency libdrm found: YES 2.4.103
Using 'PKG_CONFIG_PATH' from environment with value: ''
Run-time dependency gbm found: YES 21.0.0-devel
Using 'PKG_CONFIG_PATH' from environment with value: ''
Run-time dependency libudev found: YES 247
Configuring glmark2-es2-drm.1 using configuration
Message: Building with flavors: drm-glesv2
Build targets in project: 7

Found ninja-1.10.2 at /home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/toolchain/bin/ninja
Executing (target): ninja 
[133/136] Compiling C object src/libwsi-egl.a.p/glad_src_egl.c.o
FAILED: src/libwsi-egl.a.p/glad_src_egl.c.o 
/home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/toolchain/bin/armv8a-libreelec-linux-gnueabihf-gcc -Isrc/libwsi-egl.a.p -Isrc -I../src -I../src/glad/include -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -march=armv8-a+crc -mtune=cortex-a53 -mabi=aapcs-linux -Wno-psabi -Wa,-mno-warn-deprecated -mfloat-abi=hard -mfpu=neon-fp-armv8 -Wall -O2 -fomit-frame-pointer -fPIC -MD -MQ src/libwsi-egl.a.p/glad_src_egl.c.o -MF src/libwsi-egl.a.p/glad_src_egl.c.o.d -o src/libwsi-egl.a.p/glad_src_egl.c.o -c ../src/glad/src/egl.c
In file included from ../src/glad/include/glad/egl.h:324,
                 from ../src/glad/src/egl.c:4:
../src/glad/include/EGL/eglplatform.h:128:10: fatal error: X11/Xlib.h: No such file or directory
  128 | #include <X11/Xlib.h>
      |          ^~~~~~~~~~~~
compilation terminated.
[134/136] Compiling C++ object src/glmark2-es2-drm.p/gl-state-egl.cpp.o
FAILED: src/glmark2-es2-drm.p/gl-state-egl.cpp.o 
/home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/toolchain/bin/armv8a-libreelec-linux-gnueabihf-g++ -Isrc/glmark2-es2-drm.p -Isrc -I../src -I../src/libmatrix -I../src/glad/include -I/home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/toolchain/armv8a-libreelec-linux-gnueabihf/sysroot/usr/include -I/home/chewitt/LibreELEC.chewitt/build.LibreELEC-AMLGX.arm-9.80.8/toolchain/armv8a-libreelec-linux-gnueabihf/sysroot/usr/include/libdrm -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++14 '-DGLMARK_VERSION="2020.04"' '-DGLMARK_DATA_PATH="/usr/share/glmark2"' -march=armv8-a+crc -mtune=cortex-a53 -mabi=aapcs-linux -Wno-psabi -Wa,-mno-warn-deprecated -mfloat-abi=hard -mfpu=neon-fp-armv8 -Wall -O2 -fomit-frame-pointer -DGLMARK2_USE_EGL -DGLMARK2_USE_GLESv2 -DGLMARK2_USE_DRM -MD -MQ src/glmark2-es2-drm.p/gl-state-egl.cpp.o -MF src/glmark2-es2-drm.p/gl-state-egl.cpp.o.d -o src/glmark2-es2-drm.p/gl-state-egl.cpp.o -c ../src/gl-state-egl.cpp
In file included from ../src/glad/include/glad/egl.h:324,
                 from ../src/gl-state-egl.h:26,
                 from ../src/gl-state-egl.cpp:22:
../src/glad/include/EGL/eglplatform.h:128:10: fatal error: X11/Xlib.h: No such file or directory
  128 | #include <X11/Xlib.h>
      |          ^~~~~~~~~~~~
compilation terminated.
ninja: build stopped: subcommand failed.
```

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -24,6 +24,45 @@ jobs:
     - name: Install
       run: DESTDIR=/tmp/glmark2-install ninja -C build install
 
+  build-meson-only-drm:
+    runs-on: ubuntu-20.04
+    steps:
+    - uses: actions/checkout@v1
+    - name: Install dependencies
+      run: sudo apt install meson libdrm-dev libgbm-dev libudev-dev
+    - name: Setup
+      run: meson setup build -Dflavors=drm-gl,drm-glesv2
+    - name: Build
+      run: ninja -C build
+    - name: Install
+      run: DESTDIR=/tmp/glmark2-install ninja -C build install
+
+  build-meson-only-wayland:
+    runs-on: ubuntu-20.04
+    steps:
+    - uses: actions/checkout@v1
+    - name: Install dependencies
+      run: sudo apt install meson libwayland-dev wayland-protocols
+    - name: Setup
+      run: meson setup build -Dflavors=wayland-gl,wayland-glesv2
+    - name: Build
+      run: ninja -C build
+    - name: Install
+      run: DESTDIR=/tmp/glmark2-install ninja -C build install
+
+  build-meson-only-x11:
+    runs-on: ubuntu-20.04
+    steps:
+    - uses: actions/checkout@v1
+    - name: Install dependencies
+      run: sudo apt install meson libx11-dev
+    - name: Setup
+      run: meson setup build -Dflavors=x11-gl,x11-glesv2
+    - name: Build
+      run: ninja -C build
+    - name: Install
+      run: DESTDIR=/tmp/glmark2-install ninja -C build install
+
   build-waf:
     runs-on: ubuntu-20.04
     steps:

--- a/src/meson.build
+++ b/src/meson.build
@@ -84,7 +84,7 @@ if need_drm
     native_drm_dep = declare_dependency(
         link_with: native_drm_lib,
         dependencies: libdrm_dep,
-        compile_args: ['-DGLMARK2_USE_DRM'],
+        compile_args: ['-DGLMARK2_USE_DRM', '-D__GBM__'],
         )
 else
     native_drm_dep = declare_dependency()
@@ -119,7 +119,7 @@ if need_wayland
     native_wayland_dep = declare_dependency(
         link_with: native_wayland_lib,
         dependencies: [wayland_client_dep, wayland_egl_dep],
-        compile_args: ['-DGLMARK2_USE_WAYLAND'],
+        compile_args: ['-DGLMARK2_USE_WAYLAND', '-DWL_EGL_PLATFORM'],
         )
 else
     native_wayland_dep = declare_dependency()
@@ -201,18 +201,12 @@ else
 endif
 
 if need_egl
-    wsi_egl_lib = static_library(
-        'wsi-egl',
-        'glad/src/egl.c',
-        include_directories: include_directories('glad/include'),
-        )
     # gl-state-egl builds depend on both the GL type and the native type. To
     # avoid creating all the required combinations explicitly, make this a
     # source dependency which will be built properly as part of the final
     # glmark2 executable.
     wsi_egl_dep = declare_dependency(
-        sources: 'gl-state-egl.cpp',
-        link_with: wsi_egl_lib,
+        sources: ['gl-state-egl.cpp', 'glad/src/egl.c'],
         include_directories: [
             include_directories('glad/include'),
             include_directories('libmatrix'),
@@ -224,17 +218,11 @@ else
 endif
 
 if need_glx
-    wsi_glx_lib = static_library(
-        'wsi-glx',
-        'glad/src/glx.c',
-        include_directories: include_directories('glad/include'),
-        )
     # Although there is a single valid flavor for glx (x11-gl) and we could
     # build gl-state-glx here, make this a source dependency for convenience
     # and parity with gl-state-egl.
     wsi_glx_dep = declare_dependency(
-        sources: 'gl-state-glx.cpp',
-        link_with: wsi_glx_lib,
+        sources: ['gl-state-glx.cpp', 'glad/src/glx.c'],
         include_directories: [
             include_directories('glad/include'),
             include_directories('libmatrix'),

